protocol/chainlib/websocket_connection_limiter.go

```diff
@@ -0,0 +1,122 @@
+package chainlib
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/gofiber/websocket/v2"
+	"github.com/lavanet/lava/v4/protocol/common"
+	"github.com/lavanet/lava/v4/utils"
+)
+
+// Will limit a certain amount of connections per IP
+type WebsocketConnectionLimiter struct {
+	ipToNumberOfActiveConnections map[string]int64
+	lock                          sync.RWMutex
+}
+
+func (wcl *WebsocketConnectionLimiter) handleFiberRateLimitFlags(c *fiber.Ctx) {
+	forwardedFor := c.Get(common.IP_FORWARDING_HEADER_NAME)
+	if forwardedFor == "" {
+		// If not present, fallback to c.IP() which retrieves the real IP
+		forwardedFor = c.IP()
+	}
+	// Store the X-Forwarded-For or real IP in the context
+	c.Locals(common.IP_FORWARDING_HEADER_NAME, forwardedFor)
+
+	rateLimitString := c.Get(WebSocketRateLimitHeader)
+	rateLimit, err := strconv.ParseInt(rateLimitString, 10, 64)
+	if err != nil {
+		rateLimit = 0
+	}
+	c.Locals(WebSocketRateLimitHeader, rateLimit)
+
+	connectionLimitString := c.Get(WebSocketOpenConnectionsLimitHeader)
+	connectionLimit, err := strconv.ParseInt(connectionLimitString, 10, 64)
+	if err != nil {
+		connectionLimit = 0
+	}
+	c.Locals(WebSocketOpenConnectionsLimitHeader, connectionLimit)
+}
+
+func (wcl *WebsocketConnectionLimiter) getConnectionLimit(websocketConn *websocket.Conn) int64 {
+	connectionLimitHeaderValue, ok := websocketConn.Locals(WebSocketOpenConnectionsLimitHeader).(int64)
+	if !ok || connectionLimitHeaderValue < 0 {
+		connectionLimitHeaderValue = 0
+	}
+	// Do not allow header to overwrite flag value if its set.
+	if MaximumNumberOfParallelWebsocketConnectionsPerIp > 0 && connectionLimitHeaderValue > MaximumNumberOfParallelWebsocketConnectionsPerIp {
+		return MaximumNumberOfParallelWebsocketConnectionsPerIp
+	}
+	// Return the larger of the global limit (if set) or the header value
+	return utils.Max(MaximumNumberOfParallelWebsocketConnectionsPerIp, connectionLimitHeaderValue)
+}
+
+func (wcl *WebsocketConnectionLimiter) canOpenConnection(websocketConn *websocket.Conn) bool {
+	// Check which connection limit is higher and use that.
+	connectionLimit := wcl.getConnectionLimit(websocketConn)
+	if connectionLimit > 0 { // 0 is disabled.
+		ipForwardedInterface := websocketConn.Locals(common.IP_FORWARDING_HEADER_NAME)
+		ipForwarded, assertionSuccessful := ipForwardedInterface.(string)
+		if !assertionSuccessful {
+			ipForwarded = ""
+		}
+		ip := websocketConn.RemoteAddr().String()
+		key := wcl.getKey(ip, ipForwarded)
+		numberOfActiveConnections := wcl.addIpConnectionAndGetCurrentAmount(key)
```

Comment:
you add then immediately decrease, and not when the connection is down, no other writes to the active connections, this will always be true